### PR TITLE
fix(storage-proofs): ensure unique topological ordering of base graph

### LINKED ISSUE #744: Base DAG should have a unique topological ordering
Add edges `(i, i+1)` for all `i < N-1` for the [`parents`](https://github.com/filecoin-project/rust-fil-proofs/blob/master/storage-proofs/src/drgraph.rs#L209) of the base `Graph`. Discount that from the base degree, so we'll now have only 4 parents from bucket sampling.

Not very clear in Algorithm 1 of the [original paper](https://acmccs.github.io/papers/p1001-alwenA.pdf) but we do _know_ it has to be so since its code is based on [Argon2i](https://password-hashing.net/submissions/specs/Argon-v3.pdf).

(edit: once resolved we'll also need to update the spec which at this time is just a copy of the code here)

-------------------------------------------

PR: https://github.com/filecoin-project/rust-fil-proofs/pull/843.

### PR DESCRIPTION
**Pull request by schomatis**
BREAKING CHANGE: adding one more parent to the base graph.

Option 1 of https://github.com/filecoin-project/rust-fil-proofs/issues/744#issuecomment-519643859.

Fixes #744.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
@dignifiedquire needs to review this

**Comment by dignifiedquire** [comment]
I don't think I agree with changing the functions to return a `-1`, especially now that the constants are merged, this makes it much harder to predict how things should work.

**Comment by schomatis** [comment]
Ok, blocking for the moment then. Could you input your opinion on #744 please?

**Comment by schomatis** [comment]
Thanks for fixing the PR @DrPeterVanNostrand, LGTM.

File: storage-proofs/src/circuit/drgporep.rs [hunk 1]
```diff
@@ -566,7 +566,7 @@ mod tests {
         let sp = drgporep::SetupParams {
             drg: drgporep::DrgParams {
                 nodes,
-                degree,
+                degree: m,
```
Comment (dignifiedquire) [inline storage-proofs/src/circuit/drgporep.rs]: Either this is `m` or `degree`, but not `degree: m` please.

File: storage-proofs/src/circuit/drgporep.rs [hunk 2]
```diff
@@ -545,7 +545,7 @@ mod tests {
         let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
 
         let nodes = 12;
-        let degree = BASE_DEGREE;
+        let m = BASE_DEGREE;
```
Comment (dignifiedquire) [inline storage-proofs/src/circuit/drgporep.rs]: it should hold that `graph.degree() == BASE_DEGREE`, so the constants should be adjusted

File: storage-proofs/src/circuit/drgporep.rs (hunk 1 above)
Comment (schomatis) [reply storage-proofs/src/circuit/drgporep.rs]: Fair enough, I think Jake changed that because you didn't want the `degree - 1`.

**Comment by dignifiedquire** [comment]
merged manually into https://github.com/filecoin-project/rust-fil-proofs/pull/864

**Pull request closed without merging**